src/plugins/discover/public/application/angular/discover.js

```diff
@@ -577,18 +575,21 @@ function discoverController($element, $route, $scope, $timeout, Promise, uiCapab
     abortController = new AbortController();
 
     const searchSessionId = (() => {
-      const searchSessionIdFromURL = getSearchSessionIdFromURL(history);
+      let searchSessionIdFromURL = getSearchSessionIdFromURL(history);
```

Comment:
I'm not a huge fan of this entire block being duplicated.

What do you say we add a function called `restoreFromUrl(urlHistory)` to the session service?
The function will extract the `searchSessionId` from the URL and then apply the logic below.

Reply:
> What do you say we add a function called restoreFromUrl(urlHistory) to the session service?

This part is app-specific. It just happens that Discover and Dashboard do this similar:

It is up to the apps how they restore search sessions. Discover and Dashboard URLs structure are similar (both use `searchSessionId` query param inside hash part of the URL. 
* But not every app has a hash in the URL, so they might store `searchSessionId` as a query param in a pathname (before the hash). 
* They also might have a param named differently or be serialized in rison/json or other structures.
* We will support restoring session without the URL all together: https://github.com/elastic/kibana/issues/85126


--- 

I'd extract to common place if we'd have a shared place for Discover/Dashboard/Visualize, but as we don't, I think it is fine to leave it as is